In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
import statsmodels.api as sm


In [2]:
raw_data = pd.read_csv('BankChurners.csv')
pd.set_option('max_columns',None)
raw_data.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  Total_Relationship_Count  Months_Inactive_12_mon  \
0              39                         5                       1   
1              44                         6                       1   
2              36                         4                       1   
3              34                         3                       4   
4              21                         5                       1   

   Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0                      3       12691.0                  777          11914.0   
1                      2        8256.0                  864           7392.0   
2                      0        3418.0                    0           3418.0   
3                      1        3313.0                 2517            796.0   
4                      0        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                           0.000093                                                                                    
1                                           0.000057                                                                                    
2                                           0.000021                                                                                    
3                                           0.000134                                                                                    
4                                           0.000022                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                            0.99991                                                                                   
1                                            0.99994                                                                                   
2                                            0.99998                                                                                   
3                                            0.99987                                                                                   
4                                            0.99998

In [3]:
raw_data.describe(include = 'all')

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
count   1.012700e+04              10127  10127.000000  10127     10127.000000   
unique           NaN                  2           NaN      2              NaN   
top              NaN  Existing Customer           NaN      F              NaN   
freq             NaN               8500           NaN   5358              NaN   
mean    7.391776e+08                NaN     46.325960    NaN         2.346203   
std     3.690378e+07                NaN      8.016814    NaN         1.298908   
min     7.080821e+08                NaN     26.000000    NaN         0.000000   
25%     7.130368e+08                NaN     41.000000    NaN         1.000000   
50%     7.179264e+08                NaN     46.000000    NaN         2.000000   
75%     7.731435e+08                NaN     52.000000    NaN         3.000000   
max     8.283431e+08                NaN     73.000000    NaN         5.000000   

       Education_Level Marital_Status Income_Category Card_Category  \
count            10127          10127           10127         10127   
unique               7              4               6             4   
top           Graduate        Married  Less than $40K          Blue   
freq              3128           4687            3561          9436   
mean               NaN            NaN             NaN           NaN   
std                NaN            NaN             NaN           NaN   
min                NaN            NaN             NaN           NaN   
25%                NaN            NaN             NaN           NaN   
50%                NaN            NaN             NaN           NaN   
75%                NaN            NaN             NaN           NaN   
max                NaN            NaN             NaN           NaN   

        Months_on_book  Total_Relationship_Count  Months_Inactive_12_mon  \
count     10127.000000              10127.000000            10127.000000   
unique             NaN                       NaN                     NaN   
top                NaN                       NaN                     NaN   
freq               NaN                       NaN                     NaN   
mean         35.928409                  3.812580                2.341167   
std           7.986416                  1.554408                1.010622   
min          13.000000                  1.000000                0.000000   
25%          31.000000                  3.000000                2.000000   
50%          36.000000                  4.000000                2.000000   
75%          40.000000                  5.000000                3.000000   
max          56.000000                  6.000000                6.000000   

        Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  \
count            10127.000000  10127.000000         10127.000000   
unique                    NaN           NaN                  NaN   
top                       NaN           NaN                  NaN   
freq                      NaN           NaN                  NaN   
mean                 2.455317   8631.953698          1162.814061   
std                  1.106225   9088.776650           814.987335   
min                  0.000000   1438.300000             0.000000   
25%                  2.000000   2555.000000           359.000000   
50%                  2.000000   4549.000000          1276.000000   
75%                  3.000000  11067.500000          1784.000000   
max                  6.000000  34516.000000          2517.000000   

        Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  \
count      10127.000000          10127.000000     10127.000000   
unique              NaN                   NaN              NaN   
top                 NaN                   NaN              NaN   
freq                NaN                   NaN              NaN   
mean        7469.139637              0.759941      4404.086304   
std         9090.685324              0.219207      3397.129

In [4]:
raw_data.Income_Category.unique()

array(['$60K - $80K', 'Less than $40K', '$80K - $120K', '$40K - $60K',
       '$120K +', 'Unknown'], dtype=object)

I first feel interested in Income_Category that I think it will effect a lot on the churn rate, so I check the full category income. 
I 'm also not sure about the Attrition Flag so check it too. 

In [5]:
raw_data.Attrition_Flag.unique()

array(['Existing Customer', 'Attrited Customer'], dtype=object)

As clearly seen from the data that CLIENTNUM is not related to Attrion Customer, so delete it first

In [6]:
data = raw_data.drop('CLIENTNUM', axis=1)

In [7]:
data.isnull().sum()

Attrition_Flag                                                                                                                        0
Customer_Age                                                                                                                          0
Gender                                                                                                                                0
Dependent_count                                                                                                                       0
Education_Level                                                                                                                       0
Marital_Status                                                                                                                        0
Income_Category                                                                                                                       0
Card_Category                                   

Take a look on the Attrition_Flag to see how many percent of churn rate here.

In [8]:
data['Attrition_Flag'].value_counts(normalize = True)

Existing Customer    0.83934
Attrited Customer    0.16066
Name: Attrition_Flag, dtype: float64

In [9]:
data = data.drop(['Months_on_book','Total_Relationship_Count','Contacts_Count_12_mon'], axis = 1)

In [10]:
data = data.drop(data.loc[:,'Total_Revolving_Bal':'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2'].columns,axis = 1)


In [11]:
data['Attrition_Flag']=data['Attrition_Flag'].map({'Existing Customer':1,'Attrited Customer':0})

In [12]:
data.head()

Attrition_Flag  Customer_Age Gender  Dependent_count Education_Level  \
0               1            45      M                3     High School   
1               1            49      F                5        Graduate   
2               1            51      M                3        Graduate   
3               1            40      F                4     High School   
4               1            40      M                3      Uneducated   

  Marital_Status Income_Category Card_Category  Months_Inactive_12_mon  \
0        Married     $60K - $80K          Blue                       1   
1         Single  Less than $40K          Blue                       1   
2        Married    $80K - $120K          Blue                       1   
3        Unknown  Less than $40K          Blue                       4   
4        Married     $60K - $80K          Blue                       1   

   Credit_Limit  
0       12691.0  
1        8256.0  
2        3418.0  
3        3313.0  
4        4716.0

Create a more readerable data and convert all categorical variable to numeric

In [13]:
data['Gender']=data['Gender'].map({'M':1,'F':0})

In [14]:
data.Card_Category.unique()

array(['Blue', 'Gold', 'Silver', 'Platinum'], dtype=object)

In [15]:
data.Marital_Status.unique()

array(['Married', 'Single', 'Unknown', 'Divorced'], dtype=object)

In [16]:
data.Education_Level.unique()

array(['High School', 'Graduate', 'Uneducated', 'Unknown', 'College',
       'Post-Graduate', 'Doctorate'], dtype=object)

In [17]:
data['Marital_Status']=data['Marital_Status'].map({'Unknown':0,'Single':1,'Married':2,'Divorced':3})

In [18]:
data['Education_Level'] = data['Education_Level'].map({'Unknown':0,'Uneducated':1,'High School':2,'College':3,'Post-Graduate':4,'Graduate':5,'Doctorate':6})

In [19]:
data['Card_Category']=data['Card_Category'].map({'Blue':0,'Silver':1,'Gold':2,'Platinum':3})

In [20]:
data.head()

Attrition_Flag  Customer_Age  Gender  Dependent_count  Education_Level  \
0               1            45       1                3                2   
1               1            49       0                5                5   
2               1            51       1                3                5   
3               1            40       0                4                2   
4               1            40       1                3                1   

   Marital_Status Income_Category  Card_Category  Months_Inactive_12_mon  \
0               2     $60K - $80K              0                       1   
1               1  Less than $40K              0                       1   
2               2    $80K - $120K              0                       1   
3               0  Less than $40K              0                       4   
4               2     $60K - $80K              0                       1   

   Credit_Limit  
0       12691.0  
1        8256.0  
2        3418.0  
3        3313.0  
4        4716.0

In [21]:
data['Income_Category']=data['Income_Category'].map({'Unknown':0,'Less than $40K':1,'$40K - $60K':2,'$60K - $80K':3,'$80K - $120K':4,'$120K +':5})

In [22]:
data.head()

Attrition_Flag  Customer_Age  Gender  Dependent_count  Education_Level  \
0               1            45       1                3                2   
1               1            49       0                5                5   
2               1            51       1                3                5   
3               1            40       0                4                2   
4               1            40       1                3                1   

   Marital_Status  Income_Category  Card_Category  Months_Inactive_12_mon  \
0               2                3              0                       1   
1               1                1              0                       1   
2               2                4              0                       1   
3               0                1              0                       4   
4               2                3              0                       1   

   Credit_Limit  
0       12691.0  
1        8256.0  
2        3418.0  
3        3313.0  
4        4716.0

Declare targets and input

In [23]:
targets = data['Attrition_Flag']
inputs = data.drop(['Attrition_Flag'],axis = 1)

In [24]:
from sklearn.preprocessing import StandardScaler

Scaling as data normalization to treat the variable the same range

In [25]:
scaler = StandardScaler()
scaler.fit(inputs)
inputs_scaled = scaler.transform(inputs)

In [26]:
from sklearn.model_selection import train_test_split

Split dataset to train and test model, so that we can check the predictive potential of model

In [27]:
x_train, x_test, y_train, y_test = train_test_split(inputs_scaled, targets, test_size = 0.2, random_state=365)

In [30]:
x = sm.add_constant(x_train)
reg_log = sm.Logit(y_train,x)
results_log = reg_log.fit()

Optimization terminated successfully.
         Current function value: 0.425014
         Iterations 6


In [31]:
results_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:         Attrition_Flag   No. Observations:                 8101
Model:                          Logit   Df Residuals:                     8091
Method:                           MLE   Df Model:                            9
Date:                Fri, 25 Dec 2020   Pseudo R-squ.:                 0.02827
Time:                        15:28:34   Log-Likelihood:                -3443.0
converged:                       True   LL-Null:                       -3543.2
Covariance Type:            nonrobust   LLR p-value:                 2.836e-38
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.7250      0.032     54.153      0.000       1.663       1.787
x1            -0.0121      0.032     -0.378      0.705      -0.075       0.051
x2             0.1859      0.053      3.537      0.000       0.083       0.289
x3            -0.0657      0.031     -2.092      0.036      -0.127      -0.004
x4            -0.0323      0.031     -1.047      0.295      -0.093       0.028
x5             0.0410      0.031      1.323      0.186      -0.020       0.102
x6            -0.1639      0.054     -3.015      0.003      -0.270      -0.057
x7            -0.0537      0.036     -1.504      0.133      -0.124       0.016
x8            -0.3885      0.029    -13.171      0.000      -0.446      -0.331
x9             0.0845      0.043      1.978      0.048       0.001       0.168
==============================================================================
"""

Log-Likelihood is small and LLR p-value is extremely small show that model is significant. 
From p value, can see that x2, x6, x8 have significant on the prediction model. Age (x1) has high p-value and maybe we can delete it from the model. 
We will train model with the dataset that we already split above

In [32]:
np.set_printoptions(formatter={'float': lambda x: "{0:0.2f}".format(x)})
results_log.predict()

array([0.93, 0.85, 0.90, ..., 0.88, 0.88, 0.79])

In [34]:
np.array(data['Attrition_Flag'])

array([1, 1, 1, ..., 0, 0, 0])

In [35]:
results_log.pred_table()

array([[0.00, 1285.00],
       [0.00, 6816.00]])

In [37]:
cm_df = pd.DataFrame(results_log.pred_table())
cm_df.columns = ['Predicted 0','Predicted 1']
cm_df = cm_df.rename(index={0: 'Actual 0',1:'Actual 1'})
cm_df

Predicted 0  Predicted 1
Actual 0          0.0       1285.0
Actual 1          0.0       6816.0

In [38]:
cm = np.array(cm_df)
accuracy_train = (cm[0,0]+cm[1,1])/cm.sum()
accuracy_train

0.8413776077027527

Testing the model 

In [43]:
test_actual = y_test
test_data= sm.add_constant(x_test)

In [44]:
 def confusion_matrix(data,actual_values,model):
        pred_values = model.predict(data)
        bins=np.array([0,0.5,1])
        cm = np.histogram2d(actual_values, pred_values, bins=bins)[0]
        accuracy = (cm[0,0]+cm[1,1])/cm.sum()
        return cm, accuracy

In [45]:
# Create a confusion matrix with the test data
cm = confusion_matrix(test_data,test_actual,results_log)
cm

(array([[0.00, 342.00],
        [0.00, 1684.00]]),
 0.8311944718657454)

In [46]:
cm_df = pd.DataFrame(cm[0])
cm_df.columns = ['Predicted 0','Predicted 1']
cm_df = cm_df.rename(index={0: 'Actual 0',1:'Actual 1'})
cm_df

Predicted 0  Predicted 1
Actual 0          0.0        342.0
Actual 1          0.0       1684.0

In [47]:
# Check the missclassification rate
# Note that Accuracy + Missclassification rate = 1 = 100%
print ('Missclassification rate: '+str((1+1)/(1684+342)))

Missclassification rate: 0.0009871668311944718


The accuracy of model in predicting unchurned vs churned (with 8 variables) is 84.13%, just quite the same with percentage of existing customer.
The model predicted attrition with 0 value, which is quite weird. And I believe this is because the percentage of attrited customer in attrition Flag is too small (only ~16%) make this dataset is much more difficult. 
I will try to seperate the independent variables to analysis. 
And one 1 thing, because the right hand side of the model (I droped values from column 10 to last) so it surely effect on the result. 